### PR TITLE
ZCS 16013 - create a circle ci job for performance comparison between patch releases

### PR DESCRIPTION
**Pull request by nikdeshmukh**
A circleci job covers below steps:
1. Install software and dependencies
2. Run JMeter test on remote server
3. Generate CSV file from jtl result 
4. Run response time comparison across builds and create comparison report for soap, imap, pop and lmtp protocols
5. Copy Test files to Artifacts

### CODE REVIEW

**Comment by prashantajari** [comment]
@nikdeshmukh 
We could handle below points as future improvements:
1. In CircleCI workflow, Provide ability to run performance tests on previous build. This will be optional step and can be controlled with approval.
Job1 - Run performance tests on previous build
Job2 - Run performance tests on current build
2. We have used build1, build2 for previous and current build performance comparison. In transaction_response_times_comparison.csv files, can we capture ZCS versions instead of build1, build2.
3. We will need to archive last 4-5 build results and make use for generating performance trend reports.

Also, can we add Passed/Failed job status based on response time comparison for soap, imap, pop, lmtp?
Lets add coverage for Activesync and EWS tests.

**Comment by nikdeshmukh** [comment]
> @nikdeshmukh We could handle below points as future improvements:
> 
> 1. In CircleCI workflow, Provide ability to run performance tests on previous build. This will be optional step and can be controlled with approval.
>    Job1 - Run performance tests on previous build
>    Job2 - Run performance tests on current build
> 2. We have used build1, build2 for previous and current build performance comparison. In transaction_response_times_comparison.csv files, can we capture ZCS versions instead of build1, build2.
> 3. We will need to archive last 4-5 build results and make use for generating performance trend reports.
> 
> Also, can we add Passed/Failed job status based on response time comparison for soap, imap, pop, lmtp? Lets add coverage for Activesync and EWS tests.

**Comment by umagmrit** [comment]
@nikdeshmukh I see the directory path `/home/opc/`. Is this specific to CentOS/RHEL, or will it also work on Ubuntu and Rocky Linux?

File: regression_load_test_run.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+# Function to run JMeter test
+run_jmeter_test() {
+    test_file=$1
+    load_file=$2
+    prop_file=$3
+    results_file=$4
+    html_report=$5
+    out_files=$6
+    nohup sh /opt/apache-jmeter-5.6.2/bin/jmeter.sh -n -Juser.classpath=/home/opc/zm-load-testing/src/build/jar/zjmeter.jar -t "$test_file" -q /home/opc/zm-load-testing/config/env.prop -q "$load_file" -q "$prop_file" -l "$results_file" -e -o "$html_report" -jmeter.save.saveservice.thread_counts=true -q /home/opc/zm-load-testing/config/extras.txt > "$out_files" &
```
Comment (umagmrit) [inline regression_load_test_run.sh:25]: Can you use a variables to eliminate the repetition of the common base path wherever applicable?

**Comment by nikdeshmukh** [comment]
@umagmrit /home/opc/ is for Oracle Linux OS(Jmeter Client Machine)

**Review by umagmrit — APPROVED** [review]

File: .circleci/config.yml [hunk 2]
```diff
@@ -6,61 +6,61 @@ jobs:
...
             ssh -o StrictHostKeyChecking=no ${USER}@${SERVER}
             mkdir -p /tmp/result
           
-      # Connect to remote server and run JMeter test
+  #  Connect to remote server and run JMeter test
       - run:
           name: Run JMeter test on remote server
           command: 
             ssh -o StrictHostKeyChecking=no ${USER}@${SERVER}
             "cd /home/opc/zm-load-testing; sh regression_load_test_run.sh"
```
Comment (jiteshsojitra) [inline .circleci/config.yml:32]: /home/opc means this won't work in ubuntu?

Comment (nikdeshmukh) [reply .circleci/config.yml:32]: It won't work on Ubuntu. We are running a test on Jmeter Client which is Oracle Linux Server. This is the default server being used for regression testing.

Comment (jiteshsojitra) [reply .circleci/config.yml:32]: Okay thanks! Please fix the branch name.

Comment (nikdeshmukh) [reply .circleci/config.yml:32]: Fixed the branch name and new PR raised - https://github.com/Zimbra/zm-load-testing/pull/110

**Pull request closed without merging**